File: macos/SharedMic/Protocol/ProtocolConstants.swift
import Foundation

/// Every fixed value in protocol-v1.md, in one place.
///
/// Changing anything here is a protocol version change (protocol-v1 §1), not an
/// implementation decision.
public enum SharedMicProtocol {
    // §1 / §2
    public static let version: Int = 1
    public static let defaultPort: UInt16 = 47_800

    // §3 — envelope, big-endian
    public static let envelopeHeaderSize: Int = 5
    public static let maxPayloadBytes: Int = 1_048_576

    // §4 — audio payload. Header is big-endian; the PCM inside is little-endian.
    public static let audioHeaderSize: Int = 12
    public static let audioPCMBytes: Int = 1_920
    public static let audioPayloadSize: Int = 1_932
    public static let audioEnvelopeSize: Int = 1_937

    public static let sampleRate: Int = 48_000
    public static let channels: Int = 1
    public static let sampleFormat: String = "s16le"
    public static let samplesPerFrame: Int = 960
    public static let frameDurationUs: UInt64 = 20_000
    public static let framesPerSecond: Int = 50

    // §8 — timers
    public static let startTimeout: TimeInterval = 2.0
    public static let stopTimeout: TimeInterval = 1.0
    public static let pingInterval: TimeInterval = 15.0
    public static let peerDeadTimeout: TimeInterval = 45.0
    public static let helloDeadline: TimeInterval = 5.0

    // design spec §4.3 — reconnect
    public static let reconnectInitialDelay: TimeInterval = 0.5
    public static let reconnectMaxDelay: TimeInterval = 30.0
    public static let reconnectJitterFraction: Double = 0.2

    // §11 — pairing
    public static let tokenBytes: Int = 32
    public static let nonceBytes: Int = 32
    public static let pairingGroupSize: Int = 8
    public static let pairingStringLength: Int = 58
}

File: windows/SharedMic.Agent/Protocol/ProtocolConstants.cs
namespace SharedMic.Agent.Protocol;

/// <summary>
/// Every fixed value in protocol/protocol-v1.md, in one place. Changing any of
/// these is a protocol version change (protocol-v1.md section 1), not an
/// implementation decision.
/// </summary>
public static class ProtocolConstants
{
    public const int ProtocolVersion = 1;

    public const int DefaultPort = 47800;

    public const int EnvelopeSize = 5;
    public const int MaxPayloadBytes = 1048576;

    public const int AudioHeaderSize = 12;
    public const int PcmBytesPerFrame = 1920;
    public const int AudioPayloadSize = AudioHeaderSize + PcmBytesPerFrame;
    public const int AudioEnvelopeSize = EnvelopeSize + AudioPayloadSize;

    public const int SampleRate = 48000;
    public const int Channels = 1;
    public const string SampleFormat = "s16le";
    public const int SamplesPerFrame = 960;
    public const int FramesPerSecond = 50;
    public const int FrameDurationMs = 20;
    public const long FrameDurationUs = 20000;

    public const int AudioQueueCapacity = 25;

    public const int TokenBytes = 32;
    public const int NonceBytes = 32;
    public const int MaxAuthFailures = 5;

    public const int CertificateValidityDays = 3650;
    public const string CertificateCommonName = "shared-mic";

    public static readonly TimeSpan HelloDeadline = TimeSpan.FromSeconds(5);
    public static readonly TimeSpan HeartbeatInterval = TimeSpan.FromSeconds(15);
    public static readonly TimeSpan PeerDeadTimeout = TimeSpan.FromSeconds(45);
    public static readonly TimeSpan AuthLockoutDuration = TimeSpan.FromSeconds(30);
    public static readonly TimeSpan CertificateBackdate = TimeSpan.FromMinutes(5);
}
